torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -0,0 +1,740 @@
+    }
+  }
+
+  if (_groupDevices.find(groupId) != _groupDevices.end() &&
+      deviceList != _groupDevices[groupId]) {
+    std::string errMsg;
+    errMsg = "The current group: " + std::to_string(groupId) +
+            " has already got a GPU device list: " + _groupDevices[groupId]
+            + " associated with. Each group should only be associated with a "
+            + "given device list. Please create a new group for your provided "
+            + "device list: " + deviceList;
+    throw std::runtime_error(errMsg);
+  }
+
+  if (_ncclCommsAndEvents.find(groupId) != _ncclCommsAndEvents.end()) {
+    return std::make_pair(_ncclCommsAndEvents[groupId].first.get(),
+                          _ncclCommsAndEvents[groupId].second.get());
+  }
+
+  // Add in the device list of the group
+  _groupDevices[groupId] = deviceList;
+
+  // NCCL communication world
+  std::unique_ptr<std::vector<ncclComm_t>> comms =
+    std::unique_ptr<std::vector<ncclComm_t>>(new std::vector<ncclComm_t>());
+
+  comms->resize(input.size());
+
+  // Corresponding CUDA events
+  std::unique_ptr<std::vector<cudaEvent_t>> events =
+    std::unique_ptr<std::vector<cudaEvent_t>>(new std::vector<cudaEvent_t>());
+
+  events->resize(input.size());
+
+  // Create the unique NCCL ID and broadcast it
+  ncclUniqueId ncclId;
+  NCCL_CHECK(ncclGetUniqueId(&ncclId));
+
+  // Broadcast so that each process can have a unique NCCL ID
+  broadcastUniqueNcclId(&ncclId, &ncclId);
+
+  int curDevice = 0;
+  THCudaCheck(cudaGetDevice(&curDevice));
+
+  // Now creating the CUDA events
+  for (size_t i = 0; i < input.size(); ++i) {
+    THCudaCheck(cudaSetDevice(input[i].get_device()));
+    THCudaCheck(cudaEventCreate(&((*events)[i])));
+  }
+  // Create the communicator on each device of the input
+  NCCL_CHECK(ncclGroupStart());
+  for (size_t i = 0; i < input.size(); ++i) {
+    int nRanks = int(_numProcesses) * input.size();
+    THCudaCheck(cudaSetDevice(input[i].get_device()));
+    NCCL_CHECK(ncclCommInitRank(&((*comms)[i]),
+                                nRanks,
+                                ncclId,
+                                _rank * input.size() + i));
+  }
+  NCCL_CHECK(ncclGroupEnd());
+
+  // Restore to previous device
+  THCudaCheck(cudaSetDevice(curDevice));
+
+  // Move into the hash table
+  _ncclCommsAndEvents[groupId] = std::move(std::make_pair(std::move(comms),
+                                                          std::move(events)));
+
+  return std::make_pair(_ncclCommsAndEvents[groupId].first.get(),
+                        _ncclCommsAndEvents[groupId].second.get());
+}
+
+
+// Helper function that checks the input and output tensors for validity
+void DataChannelNccl::_tensorCheckHelper(
+    const std::vector<at::Tensor>& input,
+    const std::vector<at::Tensor>& output,
+    size_t outputOverInput) {
+
+  if (input.size() <= 0) {
+    throw std::runtime_error("Input tensor sequence cannot be empty");
+  }
```

Comment:
I don't think it should be a runtime error - this would require special-casing models with no parameters everywhere. IMO, doing something with an empty list of tensors should be a no-op rather than runtime error.